### DISCUSSION TITLE
Register resources from nuget packages in custom module

### DISCUSSION THREAD

**Discussion by PizzaConsole** [oqtane/oqtane.framework#764] [Q&A]
I was looking through this discussion: #690 but I didn't see where you were able to pull the resources directly from the nuget package into a custom Module. It appears that they downloaded the resources directly into their custom module/theme. Is that right @mikecasas ? I am specifically trying to use [ChartJs.Blazor](https://github.com/mariusmuntean/ChartJs.Blazor)

Edit 1:
I think one of the big problems I see here is this:
>In your _Host.cshtml (server-side) or in your index.html (client-side) add the following lines to the body tag after the _framework reference

As far as I understand, Oqtane loads the resources into the `<head></head>`

Edit 2:
I believe that one of the problems I am experiencing is the same as here: #727 I need to be copying the nuget .dll from the Custom Module folder to the Oqtane folder when it is packaged.

**Comment by mikecasas** [comment]
@PoisnFang Yes I download the nuget package and manually extract the assets into my module.

**Reply by PizzaConsole** [reply]
Do you have an example I could look at? I tried to do that but I couldn't get it to work...

**Reply by sbwalker** [reply]
@PoisnFang did you look at the DNF.Projects repo? It uses a Nuget package for RestSharp - and it uses Chart.js

**Reply by PizzaConsole** [reply]
I did see that it uses ChartJs I did not catch the Nuget package for RestSharp, I will take a look! I must say that the least intuitive part of the platform, for me, is integrating 3rd party resources. @sbwalker

**Reply by mikecasas** [reply]
@PoisnFang You are not alone, I think most of the questions from the community deal with that. I was hoping the docs would be up so we could write a good tutorial, but I don't know what the hold up is. In @sbwalker 's example, the JavaScript asset for chart.js comes from a CDN (see below), so there is nothing to "package" into the module, other than copying the RestSharp dll to the framework's bin (like you would with all the other project's (assets).

```
public override List<Resource> Resources => new List<Resource>()
    {
        new Resource { ResourceType = ResourceType.Stylesheet, Url = ModulePath() + "Module.css" },
        new Resource { ResourceType = ResourceType.Script, Bundle = "ChartJS", Url = "https://cdn.jsdelivr.net/npm/chart.js@2.8.0" },
        new Resource { ResourceType = ResourceType.Script, Bundle = "ChartJS", Url = ModulePath() + "Module.js" }
    };
```

**Reply by PizzaConsole** [reply]
Would there be a way to automatically extract the staticwebassets from the nuget package into the build folder?

**Reply by sbwalker** [reply]
@mikecasas its possible to use a local JavaScript file rather than a CDN... just put the JS file in the same folder as where the CSS file is located and reference it the same way ie. Url = ModulePath() + "chart.js"

**Comment by mikecasas** [comment]
@PoisnFang Here is what I do to "embed" the assets into my external module. I will use your nuget package as an example.

* I download the ChartJs.Blazor package from the nuget website
* I rename ```chartjs.blazor.1.1.0.nupkg``` to ```chartjs.blazor.1.1.0.zip```. Nuget packages are basically zip files.
* I extract the zip folder contents to another folder
* I navigate into the ```staticwebassets``` folder and copy all the js and css files needed into my external module's Server project in the wwwroot/Modules folder. If my module was 'Pets', then it would be Pets/Server/wwwroot/Modules


![assets](https://user-images.githubusercontent.com/11261164/94684005-67ffea00-02f5-11eb-9116-d06bfcb4fce7.PNG)


* In my Pets/Server project I create the following class:

```
public class ModuleStartup : IHostResources
    {
        private const string ModulePath = "/Path/To/MyModule/";

        List<Resource> IHostResources.Resources => ModuleFiles();

        private List<Resource> ModuleFiles()
        {
            List<Resource> res = new List<Resource>();
            res.Add(new Resource { ResourceType = ResourceType.Stylesheet, Url = ModulePath + "css/ChartJSBlazor.css" });
            res.Add(new Resource { ResourceType = ResourceType.Script, Url = ModulePath + "Chart.min.js" });

            return res;
        }

    }
```

You can organize the files any way you like in the Modules folder. I put my css files in a css folder...you get the idea.

Also in this class you don't have access to ModulePath, so you need to set the path somehow. Hardcoding it is not a big deal because I don't think you are going to be updating you project name and structure.

See Shaun's guidance below.

https://github.com/oqtane/oqtane.framework/discussions/690#discussioncomment-59092

Let me know if you have any questions and the above makes sense.

**Reply by PizzaConsole** [reply]
Perfect! Thank you so much! This helps a lot I will try the implementation in my project, but just in my specific Module. I don't need the Global loader.

**Reply by PizzaConsole** [reply]
I was able to get it working, using the global method, just fine. But I cannot see to get it working using the Module Method

**Reply by sbwalker** [reply]
@PoisnFang Can you share what you tried for the Module Method? I am not sure why it would not work, when it works for DNF.Projects? Is there some difference in the approach between your module and the DNF.Projects module which may explain the different in behavior?

**Reply by PizzaConsole** [reply]
@sbwalker  Below is the code on my Index page of my Custom Module, why would I get getting this error occasionally? `Error: Microsoft.JSInterop.JSException: Could not find 'ChartJsInterop' in 'window'.
Error: Could not find 'ChartJsInterop' in 'window'.`

```csharp
        public override List<Resource> Resources => new List<Resource>()
        {
            new Resource { ResourceType = ResourceType.Script, Bundle="ChartJs",  Url = ModulePath() +"ChartJsBlazor/moment-with-locales.min.js" },
            new Resource { ResourceType = ResourceType.Script,Bundle="ChartJs", Url = ModulePath() +"ChartJsBlazor/Chart.min.js" },
            new Resource { ResourceType = ResourceType.Script,Bundle="ChartJs", Url = ModulePath() +"ChartJsBlazor/ChartJsBlazorInterop.js" },
            new Resource { ResourceType = ResourceType.Stylesheet,  Url = ModulePath() +"ChartJsBlazor/ChartJSBlazor.css" }
        };
```

**Reply by mikecasas** [reply]
Do you have a public repo for this?  What errors are you getting?

**Reply by PizzaConsole** [reply]
See above ^

**Reply by mikecasas** [reply]
@PoisnFang Basically your component is rendering before the JS files.

**Reply by PizzaConsole** [reply]
That's what I figured, but How would you prevent that?

**Reply by mikecasas** [reply]
I need to look back at the threads.

**Reply by mikecasas** [reply]
My "solution" was to use IHostResources, or the "global" method above.

**Reply by PizzaConsole** [reply]
Yes, that works for me too, but in order to follow the the Oqtane philosophy this will need to get addressed. I am also using [BlazorInputFile](https://github.com/SteveSandersonMS/BlazorInputFile) `new Resource { ResourceType = ResourceType.Script,  Url = ModulePath()+"BlazorInput/inputfile.js" },` in the Index Page and it is throwing the error Every single time.
`Error: Microsoft.JSInterop.JSException: Could not find 'BlazorInputFile' in 'window'.
Error: Could not find 'BlazorInputFile' in 'window'.`
I will make an issue for it.

**Reply by sbwalker** [reply]
For chart.js, are you sure that you are loading the JS files in the correct order. The bundle property means they need to be grouped and loaded in the order they are declared.  In addition it would be interesting to know what the JSInterop class looks like which calls the methods in these JS files, as its possible the issue is in those wrappers.

**Reply by sbwalker** [reply]
Also it would helpful to know if the JS files are in the optimal format... it appears they may not be using an appropriate namespace.

**Reply by sbwalker** [reply]
Is there a reason why you would use the BlazorInputFile when the Oqtane FileManager component already serves this purpose?

**Reply by PizzaConsole** [reply]
The order of the ChartJs.Blazor follows the installation instructions. Like I mentioned, it only Sometimes throws the error, whereas BlazorInputFile Always throws it.  I found the BlazorInputFile easy to implement because I already had experience with it.

**Reply by sbwalker** [reply]
Loading JavaScript on-demand in a SPA framework based on the specific needs of a page is really complicated. Microsoft avoided this for the most part in their simplistic Blazor templates by requiring all JavaScript to be declared in the _Host.cshtml - however this is really inefficient in a large enterprise application as you are sending a ton of JavaScript to the browser even if the specific user does not require it ( ie. JavaScript related to admin areas of the application are only applicable to a tiny subset of users ). We spent a lot of time and effort in Oqtane to come up with the current on-demand JavaScript loading solution, so it would be useful to try and figure out why it works in some cases but not others. Obviously Oqtane still has IHostResources as a fallback - but again, that solution means that you are always sending JavaScript to the browser - it is no longer on-demand based on the needs of the user/page.

**Reply by sbwalker** [reply]
@PoisnFang I have looked into the problem you were experiencing and I can explain the issue. In V1.0.1 @jimspillane and I spent a lot of time coming up with a lazy-loading/on-demand JavaScript loading solution which handles dependencies between JS files, etc... What we discovered as part of that exercise is that the JavaScript needs to be loaded in the component where it is referenced. The reason for this is because Blazor components are completely asynchronous and independent from one another, so if you try to rely on dependencies between components you can run into race conditions which are very difficult to diagnose because they will occur randomly ( in reality it is not random - it is based on performance characteristics of the run-time environment at a specific moment in time - but it is nearly impossible to reproduce those characteristics consistently ). 

In your Oqtane module, you are using a wrapper component for ChartJSBlazor - and that wrapper component is not taking responsibility for loading the JavaScript itself. Instead, your module component is declaring/loading the JavaScript and ChartJSBlazor is trying to render the graph. So you are relying on a dependency between your module component and ChartJSBlazor - and ChartJSBlazor is executing before your module component is finished its work ( ie. a race condition ) - so it is throwing an error. If you compare this to the DNF.Projects module you will notice its usage of Chart.js works fine because the JavaScript is declared and loaded in the same module component as where it is used for displaying the graph - no chance for a race condition. Your other example of BlazorInputFile has the same problem. BlazorInputFile is a wrapper around some JavaScript and it does not take responsibility for loading the JavaScript itself. 

So this explains the problem. However taking a step back, there is a bigger question. The current Oqtane solution for JavaScript loading was designed with performance and efficiency in mind ( ie. as per the Philosophy ). The goal was to avoid sending a ton of JavaScript to the client browser unless that user actually needs it for their user experience. This approach is contrary to the vast majority of SPA applications today which forcefully send a ton of JavaScript to the client browser whether they actually need it or not. And the current Oqtane solutions works well - as long as your component is taking full responsibility for loading and using the JavaScript.

That being said, I am seeing a lot of Blazor components emerge which are wrappers around JavaScript - both from independent developers as well as from Microsoft itself. Due to the fact that Blazor does not have a solution like Oqtane where a component can be fully encapsulated and load its own JavaScript resources, component authors usually provide documentation instructions to manually add the resources to your _host.cshtml file. This is a very poor integration model for a variety of reasons which should be obvious - but the most important thing is that it works. It works because on startup it is forcing the browser to download and load ALL of the JavaScript resources for the entire application prior to executing any Blazor components.

So at this point, although the Oqtane method is superior in terms of efficiency and performance, I am questioning if we should not alter our approach in order to make integrations simpler. We have already demonstrated through the IHostResources interface that we can dynamically inject JavaScript resources in _host.cshtml. However the implementation could be simplified so that the JavaScript resources declared within a module ( ie. the exact way we expect developers to register them today ) are injected into _host.cshtml on startup rather than dynamically injected on-demand using JSInterop. As part of this change I would suggest we deprecate IHostResources as there is no point in having 2 solutions for the same purpose - it just leads to confusion as it is not clear to developers which one they should use. Overall this approach would make module development more intuitive and reliable for developers, and would probably help avoid a lot of troubleshooting and explanation. 

I would like to hear @jimspillane comment on this, as we spent a LOT of time coming up with the current solution.

**Reply by mikecasas** [reply]
Maybe not deprecate it, but allow the two different ways to load the JS, since one provides a better performance. But I agree the default pattern should be to load the JS in the "host" file, so it is easier for everyone. I know it can be confusing, but I would not want to lose the "superior" nature of Oqtane. Just my 2 cents.

**Reply by sbwalker** [reply]
@jimspillane in order to support both approaches perhaps we could enhance the Resource model to have a bool property for LazyLoad. The default would be false. Then module developers can still declare their resources in a consistent manner within their module components - and the framework can deal with how to inject them behind the scenes.

**Reply by jimspillane** [reply]
@sbwalker I was thinking about that the other day.  Are we going to migrate to .NET 5? Doesn't the [Blazor JavaScript isolation and object references](https://devblogs.microsoft.com/aspnet/asp-net-core-updates-in-net-5-release-candidate-1/#blazor-javascript-isolation-and-object-references) feature resolve many of the issues that we experiencing?

Hers is a demo [Blazor .NET 5 - JavaScript Isolation - JavaScript Files per Component (Finally!)](https://www.youtube.com/watch?v=rmkYXkVzIck)

**Reply by sbwalker** [reply]
@jimspillane yes the plan is to migrate to .NET 5 ( we actually want to have a release ready to coincide with the official release on Nov 10 ). That being said, the initial goal is purely compatibility - ensuring the current feature set works on .NET 5. The next step would be to determine if/how any of the new Blazor features can be integrated. 

I looked at the minimal information about JavaScript isolation and it appears to be a dynamic solution but it raises a lot of questions for me:

- it appears that in order to use it, you need to specify your Javascript as modules - which means that most JavaScript that people are using today in Blazor will need to be modified/rewritten in order to use this new feature
- there is a new magic string for "import" which I assume is the loader - however nowhere does it mention if the import must be in the same component where the Javascript is used. I assume so based on the async nature of components and what we discovered already in our experiments - but it would be nice if there was more technical guidance available to developers. 
- I have explored the .NET Core source on Github to try and figure out how "import" works under the covers but I have not been able to find the source files. Specifically I am interested in how they are dealing with JavaScript dependencies - as I do not see any support for "bundling" - which was a critical issue we experienced/solved in Oqtane by using LoadJS
- it appears that JS Interop wrapper classes also need to be rewritten to use the new JSObjectReference feature
- I also do not how it supports JavaScript resources which are remote ( ie. on a CDN ) and have Integrity keys

And from a higher level it appears that in .NET 5 there will be multiple methods for doing JSInterop - as Microsoft will still need to support the current methods - and each method has its own unique implementation patterns. And it will take some time before the .NET ecosystem adapts to the new methods... especially if you consider that if someone has built a component today and they want to support both .NET Core 3.1 (LTS) and .NET 5 (GA) then they will either need to create/support 2 versions of their component - or what is more likely, choose to manage a single code base by writing to the lowest common denominator, which means using the current methods rather than the .NET 5 methods. Essentially what this means from an Oqtane perspective is that we will need to support both methods for module developers.

**Reply by jimspillane** [reply]
> * it appears that in order to use it, you need to specify your Javascript as modules - which means that most JavaScript that people are using today in Blazor will need to be modified/rewritten in order to use this new feature

Correct.  In many cases I think adding _export_ to the functions will be the only modification needed.
 
> * there is a new magic string for "import" which I assume is the loader - however nowhere does it mention if the import must be in the same component where the Javascript is used. I assume so based on the async nature of components and what we discovered already in our experiments - but it would be nice if there was more technical guidance available to developers.

I did not see much guidance available as well.  From what I can tell, controlling the scope to the module level is the benefit. 

> * I have explored the .NET Core source on Github to try and figure out how "import" works under the covers but I have not been able to find the source files. Specifically I am interested in how they are dealing with JavaScript dependencies - as I do not see any support for "bundling" - which was a critical issue we experienced/solved in Oqtane by using LoadJS

I found the _import_ referenced here  [Microsoft.JSInterop.ts](https://github.com/dotnet/aspnetcore/blob/release/5.0/src/JSInterop/Microsoft.JSInterop.JS/src/src/Microsoft.JSInterop.ts#L58-L67) but I am not sure how it works.

> * it appears that JS Interop wrapper classes also need to be rewritten to use the new JSObjectReference feature
> * I also do not how it supports JavaScript resources which are remote ( ie. on a CDN ) and have Integrity keys

I created a small test using Chart.js.  I created a local script file that imports from the CDN.
https://github.com/jimspillane/BlazorAppChartJS/blob/master/BlazorAppChartJS/wwwroot/js/charttest.js

Calling it is rather simple.
https://github.com/jimspillane/BlazorAppChartJS/blob/master/BlazorAppChartJS/Pages/Index.razor

From what I can tell, the scope is limited to the life of the  _module variable and no global namespace was used.

> And from a higher level it appears that in .NET 5 there will be multiple methods for doing JSInterop - as Microsoft will still need to support the current methods - and each method has its own unique implementation patterns. And it will take some time before the .NET ecosystem adapts to the new methods... especially if you consider that if someone has built a component today and they want to support both .NET Core 3.1 (LTS) and .NET 5 (GA) then they will either need to create/support 2 versions of their component - or what is more likely, choose to manage a single code base by writing to the lowest common denominator, which means using the current methods rather than the .NET 5 methods. Essentially what this means from an Oqtane perspective is that we will need to support both methods for module developers.

Agreed.  If the new isolation feature requires ES6 modules, supporting both methods will be wise.  I had to use the Chart.js 3.0 beta because the 2.0 version is not an ES module. I imagine we might run into other popular libraries that don't support modules yet.

**Reply by sbwalker** [reply]
@jimspillane thank you for finding the code on Github. I don’t understand it either... but I am guessing that if you compared it to the internals of LoadJS you may see some similarities. One thing that does jump out at me though is it relies on “./“ notation to identify resources in wwwroot - which is a concept which Oqtane probably needs to support as well referencing static resources.

**Reply by jimspillane** [reply]
[Dynamic Imports](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#Dynamic_Imports) come with some benefits that are similar to LoadJS.  Unlike the LoadJS solution, the ES modules are _deleted_ when disposed of rather than loitering in the global namespace.

**Comment by sbwalker** [comment]
In the majority of cases you should not need the IHostResources interface. It is only available if you need to include resources that will be used across your entire site ( ie. such as a UI component library which you are using in every module ). Typically you should access your resources using the method in the DNF.Projects example repo. I actually published a video series a while ago where I used DNF.Projects as an example - and I included instructions on adding Nuget packages and static asset references https://www.youtube.com/watch?v=JPfUZPlRRCE&list=PLYhXmd7yV0elLNLfQwZBUlM7ZSMYPTZ_f

The main thing you need to keep in mind when building external modules is that you need to deploy them as well as all of their dependencies into the Oqtane framework using the specific folder conventions it requires. So you can pull Nuget packages into your module just like any other .NET application... but you need to ensure all of the contents of the Nuget package are deployed to Oqtane. An easy way to do this is to do a mock Publish from Visual Studio to a folder and locate all of the DLLs and static assets - these will need to be deployed to your Oqtane instance.

![image](https://user-images.githubusercontent.com/4840590/94685993-2c1a5400-02f8-11eb-8937-6e1097acb3a4.png)

**Reply by PizzaConsole** [reply]
> The main thing you need to keep in mind when building external modules is that you need to deploy them as well as all of their dependencies into the Oqtane framework using the specific folder conventions it requires. 

Right and that's where the platform loses intuitiveness. You have to **Manually** make this happen, whereas in just a plain Blazor project you add in the Nuget package and it just "magically" works. It adds in a few extra steps and some developers might not be used to doing.

**Reply by sbwalker** [reply]
You are correct. In a plain Blazor project where everything is tightly coupled and statically linked, the tooling can do some magic for you so that you are not aware of what is happening under the covers. In a loosely coupled dynamic application, the tooling cannot do that magic and you need to manage it yourself.

**Reply by PizzaConsole** [reply]
Unfortunately it leaves a lot of room for human error that can be very time costly.

**Reply by sbwalker** [reply]
There is a saying "with greater power comes greater responsibility" - this is definitely true for more advanced software frameworks

**Comment by PizzaConsole** [comment]
@sbwalker Just a thought but, shouldn't `<CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>`
be included by default in the Module Creation Template? It is necessary in order to copy the 3rd party resources over.

**Reply by sbwalker** [reply]
Yes, it could definitely be added to the external template to make it simpler

**Reply by PizzaConsole** [reply]
I will make a PR

**Comment by sbwalker** [comment]
@PoisnFang in #829 I added enhancements based on the conversation in this thread. It is now possible to declare Resources in your module or theme and specify a property Declaration of "Global". By setting this property, the framework will inject the resource into the _host.cshtml as a global resource. This should make it simpler for you to use components like ChartJSBlazor.

@markdav-is the changes in #829 also support a new property for Location where you can specify Head or Body. Specifying a Declaration of Global and a Location of Body will instruct the framework to add the script reference to the end of the body in _host.cshtml

@mikecasas note that IHostedResources is still available and supports the enhancements mentioned above.

This should not be a breaking change. It will need some thorough testing.

**Reply by PizzaConsole** [reply]
@sbwalker Very excellent!

**Reply by sbwalker** [reply]
@PoisnFang @markdav-is do you think you will have time to test this new feature with your 2 use cases you described above?

**Reply by PizzaConsole** [reply]
@sbwalker My Module is currently running the v1.0.4 is this a change what will only be available in the v2 release?

**Reply by sbwalker** [reply]
Currently yes that is the plan. All enhancements in master will be part of the 2.0 release on Nov 10

**Reply by PizzaConsole** [reply]
I will have to try and find sometime to rearrange my project to test it with the latest changes then, or I will just have to wait until v2 comes out to swap it
